### ISSUE TITLE
Sharding open questions and todos remaining

### ISSUE THREAD

**Issue by lkatalin** [sigstore/rekor#711]
@priyawadhwa and I wanted a place to record some open questions, bugs, and todo items remaining for the sharding work.

### Questions
- [x] Is it expected that `rekor-cli search` will return artifacts from all available shards in that Rekor instance? (We think yes)
- [x] How will clients interact with sharded rekor instances when verifying? (Each shard will have its own signing key(s) according to the initial design doc) (Issue: #721) What is the impact of this? (@asraa has been working on this from the cosign side, but currently it seems there are not todo items relating to this in Rekor)
- [x] How can we "remove" an entry from a log? We don't currently have logs pointing at other logs in a linked list format. (deferred until after GA)

### Todos
- [x] Get Public Key via the API per Tree ID for loginfo/verification in general (PR: #734 and PR: #733)
- [x] Implement way to pass in inactive shard map as a config file (PR: #742)

### Bugs
- [x] Need to return universal index over shards to user upon artifact upload (PR: #725 )
- [x] Need to correctly update `state.json` when new shard is created (Issue: #713 )
- [x] Return TreeID as string in loginfo so that it doesn't get rounded down in jq (PR: #712 )
- [x] Leaving `tlog_id` unspecified or as `0` seems to default to old (inactive) shard on server startup. (Issue: #726, PR: #727)

### Documentation To Dos
- [x] Neglecting to populate a shard config file and pass it in at server startup breaks index resolution when manually restarting the local server. This is because state is saved in the mariadb backend and a new shard is created at each startup, so re-starting multiple times creates multiple shards.  (#749 ) --> this would also be fixed by building up the `logRangeMap` internally. UPDATE: This usage is actually an antipattern, so we will cover this in documentation.
- [x] Add docs on sharding, including how to populate the sharding config file (PR https://github.com/sigstore/sigstore-website/pull/145)

### Expected bugs
- Using the `v0.5` client with a sever that returns `entryIDs` (aka server code after #623 is merged) will work correctly for upload and GET requests for the active tree, but will not work correctly with GET requests for archived trees. This is because the `v0.5` client does not yet have correct tree routing to pass to the server. (Fixed by #750 )

### Lower-priority todos (can be deferred to after Sigstore GA)
- [ ] Update terminology throughout Rekor to refer to `treeIDs` and `entryIDs` #625 #672 
- [x] Update `createAndInitTree()` debug output to reflect use of pre-existing tree when appropriate (PR: #724 )
- [ ] Check `logRangeMap` passed in by user against actual shards in trillian (the shard config should be a (strict or not strict) subset of trees that exist in the trillian backend for that instance) (Issue #770 )

### Rejected changes:
- Add a way to build up the `logRangeMap` internally by querying trillian to avoid user error when passing it on the CLI

More to come!

**Comment by lkatalin** [comment]
Potentially we could update `rekor-cli search` to **optionally** only search a specific shard, if that would be useful?

**Comment by priyawadhwa** [comment]
TODO: we'll have to be able to get Public Key via the API per Tree ID for loginfo/verification in general

**Comment by lkatalin** [comment]
> TODO: we'll have to be able to get Public Key via the API per Tree ID for loginfo/verification in general

Thanks, I added this to the list up top.

**Comment by lkatalin** [comment]
All the critical tasks for GA are done! Closing now.
